File: src/titanium.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

/*
WARNING -- dangerous compilation
gcc -fno-stack-protector -o titanium titanium.c
*/

int main(int argc, char *argv[]) {
  char* flag = getenv("FLAG");
  char* secret = getenv("SECRET_KEY");
  struct ab { 
  	char b[8];
    char a[32];   
  } s;
  if(argc < 2) {
    return 1;
  }
  strcpy(s.a, secret);
  sscanf(argv[1], "%s", s.b);
  if(strcmp(s.a, secret) != 0) {
    printf("%s", flag);
  } else {
    return 1;
  }
  return 0;
}

File: src/titanium.py
from flask import Flask, request
import subprocess
import os

app = Flask(__name__)

@app.route('/')
def index():
    os.system('gcc -fno-stack-protector -o titanium titanium.c')
    return 'api up'

# Route for 'titanium' buffer overflow challenge
@app.route('/titanium')
def titanium():
  flag = os.environ['FLAG']
  os.system(f'export FLAG={flag}')
  secret_key = os.environ['SECRET_KEY']
  os.system(f'export SECRET_KEY={secret_key}')
  arg1 = request.args.get('arg1')
  try:
    output = subprocess.check_output(['./titanium', arg1])
    return output.decode('utf-8')
  except:
    return ''

if __name__ == '__main__':
    app.run(host='0.0.0.0')
